### PR TITLE
tests: use channel in block_engine

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,all,
     This PR is used to fix https://github.com/pingcap/tikv/issues/1721
      It use channel in `BlockEngine`.  when `async_snapshot` or `async_write` is finished, it send a message to main thread before blocked.
      @disksing @zhangjinpeng1987 @BusyJay  PTAL

### CODE REVIEW

**Comment by disksing** [comment]
LGTM.

File: tests/storage/test_storage.rs [hunk 1]
```diff
@@ -675,12 +675,12 @@ fn bench_txn_store_rocksdb_put_x100(b: &mut Bencher) {
...
 
 fn test_storage_1gc_with_engine(engine: Box<Engine>, ctx: Context) {
-    let engine = util::BlockEngine::new(engine);
+    let mut engine = util::BlockEngine::new(engine);
     let config = Config::default();
     let mut storage = Storage::from_engine(engine.clone(), &config).unwrap();
     storage.start(&config).unwrap();
-
-    engine.block_snapshot();
+    let (stx, srx): (Sender<bool>, Receiver<bool>) = channel();
```
Comment (siddontang) [inline tests/storage/test_storage.rs]: need declaring type here?

File: tests/storage/test_raft_storage.rs [hunk 2]
```diff
@@ -180,8 +176,8 @@ fn test_scheduler_leader_change_twice() {
             tx.send(1).unwrap();
         })
         .unwrap();
-    // Sleep a while, the prewrite should be blocked at snapshot stage.
-    thread::sleep(Duration::from_millis(200));
+    // wait for the message, the prewrite should be blocked at snapshot stage.
+    srx.recv().unwrap();
```
Comment (siddontang) [inline tests/storage/test_raft_storage.rs]: need to use recv_timeout to avoid wait forever?

File: tests/storage/util.rs [hunk 3]
```diff
@@ -38,35 +41,53 @@ impl BlockEngine {
...
-    pub fn unblock_write(&self) {
+    pub fn unblock_write(&mut self) {
         self.block_write.store(false, Ordering::SeqCst);
+        self.set_sender(None);
     }
 
-    pub fn block_snapshot(&self) {
+    pub fn block_snapshot(&mut self, sender: Sender<bool>) {
         self.block_snapshot.store(true, Ordering::SeqCst);
+        let mut data = self.sender.lock().unwrap();
```
Comment (siddontang) [inline tests/storage/util.rs]: use set_sender instead

File: tests/storage/util.rs [hunk 4]
```diff
@@ -38,35 +41,53 @@ impl BlockEngine {
...
 
 impl Engine for BlockEngine {
     fn async_write(&self, ctx: &Context, batch: Vec<Modify>, callback: Callback<()>) -> Result<()> {
         let block_write = self.block_write.clone();
+        let sender = self.sender.clone();
         self.engine.async_write(ctx,
                                 batch,
                                 box move |res| {
             thread::spawn(move || {
+                if block_write.load(Ordering::SeqCst) {
```
Comment (siddontang) [inline tests/storage/util.rs]: similar code with L105?

File: tests/storage/test_storage.rs (hunk 1 above)
Comment (AndreMouche) [inline tests/storage/test_storage.rs]: It's just used to be more clear for reviewer. @disksing

File: tests/storage/util.rs (hunk 4 above)
Comment (AndreMouche) [inline tests/storage/util.rs]: yes,`block_write ` and  `block_snapshot ` share one `sender` now. @siddontang

File: tests/storage/test_storage.rs (hunk 1 above)
Comment (siddontang) [inline tests/storage/test_storage.rs]: it is redundant here, we can know the type exactly when we use it.

File: tests/storage/util.rs (hunk 4 above)
Comment (siddontang) [inline tests/storage/util.rs]: so you should extract to a function

**Comment by AndreMouche** [comment]
PTAL

File: tests/storage/util.rs [hunk 5]
```diff
@@ -38,35 +41,58 @@ impl BlockEngine {
...
 
-    pub fn unblock_snapshot(&self) {
+    pub fn unblock_snapshot(&mut self) {
         self.block_snapshot.store(false, Ordering::SeqCst);
+        self.set_sender(None);
+    }
+}
+
+// try_to_notify tries to send message when block status = true && sender is not none.
+fn try_to_notify(block: Arc<AtomicBool>, sender: Arc<Mutex<Option<Sender<bool>>>>) {
```
Comment (siddontang) [inline tests/storage/util.rs]: try_notify

**Pull request merged**